src/EditorFeatures/Core/Implementation/Classification/SemanticClassificationTaggerProvider.cs

```diff
@@ -38,45 +56,135 @@ internal partial class SemanticClassificationTaggerProvider : AbstractAsynchrono
+        public override ITaggerEventSource CreateEventSource(ITextView textViewOpt, ITextBuffer subjectBuffer)
+        {
+            // Note: we don't listen for OnTextChanged.  Text changes to this this buffer will get
+            // reported by OnSemanticChanged.
+            return TaggerEventSources.Compose(
+                TaggerEventSources.OnSemanticChanged(subjectBuffer, TaggerDelay.Short, _semanticChangeNotificationService),
+                TaggerEventSources.OnDocumentActiveContextChanged(subjectBuffer, TaggerDelay.Short));
+        }
 
-        protected override IEnumerable<Option<bool>> TagSourceOptions
+        public override async Task ProduceTagsAsync(TaggerContext<IClassificationTag> context)
         {
-            get
+            Debug.Assert(context.SpansToTag.IsSingle());
+            Debug.Assert(context.CaretPosition == null);
+
+            var spanToTag = context.SpansToTag.Single();
+            var document = spanToTag.Document;
+
+            if (document == null)
+            {
+                return;
+            }
+
+            if (_classificationService == null)
             {
-                yield return InternalFeatureOnOffOptions.SemanticColorizer;
+                _classificationService = document.Project.LanguageServices.GetService<IEditorClassificationService>();
             }
+
+            if (_classificationService == null)
+            {
+                return;
+            }
+
+            var cancellationToken = context.CancellationToken;
+            await ProduceTagsAsync(context, spanToTag).ConfigureAwait(false);
         }
 
-        protected override ProducerPopulatedTagSource<IClassificationTag> CreateTagSourceCore(ITextView textViewOpt, ITextBuffer subjectBuffer)
+        private async Task ProduceTagsAsync(TaggerContext<IClassificationTag> context, DocumentSnapshotSpan spanToTag)
         {
-            return new SemanticBufferTagSource<IClassificationTag>(
-                subjectBuffer,
-                CreateTagProducer(),
-                CreateEventSource(textViewOpt, subjectBuffer),
-                AsyncListener,
-                NotificationService,
-                this.RemoveTagsThatIntersectEdits,
-                this.SpanTrackingMode);
+            if (await TryClassifyContainingMemberSpan(context, spanToTag).ConfigureAwait(false))
+            {
+                return;
+            }
+
+            // We weren't able to use our specialized codepaths for semantic classifying. 
+            // Fall back to classifying the full span that was asked for.
+            await ClassifySpansAsync(context, spanToTag).ConfigureAwait(false);
         }
 
-        protected override ITagProducer<IClassificationTag> CreateTagProducer()
+        private async Task<bool> TryClassifyContainingMemberSpan(TaggerContext<IClassificationTag> context, DocumentSnapshotSpan spanToTag)
         {
-            return new TagProducer(_typeMap);
+            var range = context.TextChangeRange;
```

Comment:
Yes.  You can definitely get a null range.  The changes represent the set of text changes since the last tagging fully completed.  If no text changed, you'll get null.   I can go into more detail if you need.